cpp/src/merge/merge.cu

```diff
@@ -0,0 +1,452 @@
+  BitMaskT const* const __restrict__ source_right_mask= right_dcol.null_mask();
+  BitMaskT* const destination_mask = out_dcol.null_mask();
+  
+  auto active_threads =
+    __ballot_sync(0xffffffff, destination_row < num_destination_rows);
+
+  while (destination_row < num_destination_rows) {
+    index_type const& merged_idx = merged_indices[destination_row];
+    side const src_side = thrust::get<0>(merged_idx);
+    cudf::size_type const src_row  = thrust::get<1>(merged_idx);
+    bool const from_left{src_side == side::LEFT};
+    bool source_bit_is_valid{true};
+    if (left_have_valids && from_left) {
+      source_bit_is_valid = left_dcol.is_valid_nocheck(src_row);
+    }
+    else if (right_have_valids && !from_left) {
+      source_bit_is_valid = right_dcol.is_valid_nocheck(src_row);
+    }
+
+    // Use ballot to find all valid bits in this warp and create the output
+    // bitmask element
+    BitMaskT const result_mask{
+      __ballot_sync(active_threads, source_bit_is_valid)};
+
+    cudf::size_type const output_element = cudf::word_index(destination_row);
+
+    // Only one thread writes output
+    if (0 == threadIdx.x % warpSize) {
+      destination_mask[output_element] = result_mask;
+    }
+
+    destination_row += blockDim.x * gridDim.x;
+    active_threads =
+      __ballot_sync(active_threads, destination_row < num_destination_rows);
+  }
+}
+
+void materialize_bitmask(cudf::column_view const& left_col,
+                         cudf::column_view const& right_col,
+                         cudf::mutable_column_view& out_col,
+                         index_type const* merged_indices,
+                         cudaStream_t stream) {
+  constexpr cudf::size_type BLOCK_SIZE{256};
+  cudf::util::cuda::grid_config_1d grid_config {out_col.size(), BLOCK_SIZE };
+
+  auto p_left_dcol  = cudf::column_device_view::create(left_col);
+  auto p_right_dcol = cudf::column_device_view::create(right_col);
+  auto p_out_dcol   = cudf::mutable_column_device_view::create(out_col);
+
+  auto left_valid  = *p_left_dcol;
+  auto right_valid = *p_right_dcol;
+  auto out_valid   = *p_out_dcol;
+  
+  if (p_left_dcol) {
+    if (p_right_dcol) {
+      materialize_merged_bitmask_kernel<true, true>
+        <<<grid_config.num_blocks, grid_config.num_threads_per_block, 0, stream>>>
+        (left_valid, right_valid, out_valid, out_col.size(), merged_indices);
+    } else {
+      materialize_merged_bitmask_kernel<true, false>
+        <<<grid_config.num_blocks, grid_config.num_threads_per_block, 0, stream>>>
+        (left_valid, right_valid, out_valid, out_col.size(), merged_indices);
+    }
+  } else {
+    if (p_right_dcol) {
+      materialize_merged_bitmask_kernel<false, true>
+        <<<grid_config.num_blocks, grid_config.num_threads_per_block, 0, stream>>>
+        (left_valid, right_valid, out_valid, out_col.size(), merged_indices);
+    } else {
+      materialize_merged_bitmask_kernel<false, false>
```

Comment:
Technically in this code path, as an optimization you could simply do a memset on the output bitmask to set all the bits.

Reply:
Okay. I'll look into that.

Reply:
The legacy code had a different semantics: `left_valid`, `right_valid` were set to true/false depending on whether the old `valid` pointer was non-nullptr/nullptr, respectively). 

So, just to confirm: the `<false, false>` means _no_ valid pointer, which is equivalent to ALL_VALID; hence the memset for this branch should be to ALL_VALID. Correct?

Reply:
Correct.

Reply:
@jrhemstad I've added a `set_null_mask()` in `null_mask.hpp/cu` and the way it's being used in `ColumnMerge` addressed the optimization requirement above. 

In fact, the case `materialize_merged_bitmask_kernel<false, false>()` will not be called anymore because in `ColumnMerger` `materialize_bitmask()` will only be called if `(lcol.has_nulls() || rcol.has_nulls()` (which excludes the `<false, false>` case above). 

And `merged_view` gets its `null_mask` preset to `ALL_VALID` (which is what  `materialize_merged_bitmask_kernel<false, false>()` would have done, anyway).

Please let me know if that's acceptable.

Reply:
I would eliminate the call to `materialize_merged_bitmask_kernel<false, false>`.